projecttacoma/fqm-testify
# Highlighting update

# Summary
The measure highlighting for individual patients now automatically updates when the patient data changes rather than the user having to click the calculate button each time.

## New Behavior
- There is no more calculate button in `MeasureHighlightingPanel.tsx` because the measure highlighting automatically recalculates when there is a change to the patient data. Instead, there is an loading icon that shows when the measure highlighting is being recalculated, or a check mark that shows that the highlighting is up to date. 
- When a user selects a patient, the measure highlighting for that patient immediately appears, even if the patient does not have any resources. When a user adds resources, the highlighting updates to reflect those changes. The highlighting also updates when the patient data is edited, when resources are added or edited, and when resources are deleted.
- Since the measure highlighting also changes when the measure bundle or measure period changes, the measureReport calculation is reran every time the user navigates to the TestCaseEditorPage. This is how the measure highlighting initially shows up and how the measure report is recalculated if the user navigates back to the measure upload page and uploads a different bundle or selects a different measurement period.

## Code Changes
- `MeasureCalculation.tsx` - this helper function/file was created to hold a function that takes in a patientTestCase, a measure bundle, and a measurement period start and end and returns a promise for a MeasureReport. This is in place of the function that formerly resided in `MeasureHighlightingPanel.tsx`.
- `MeasureHighlightingPanel.tsx` - the calculate function was removed and replaced with a call to the `calculateMeasureReport` helper function and the calculate button was also removed. Icons were added to the panel to indicate when the panel was loading or up-to-date.
- `PatientCreationPanel.tsx` - calls to the `calculateMeasureReport` helper function are added where the patient data changes. 
- `ResourceDisplay.tsx` - calls to the calculateMeasureReport` helper function are added where the patient data changes. 
- `generate-test-cases.tsx`- a function is added and called in order to recalculate the each patients' measure reports when the user navigates to the `TestCaseEditorPage`. Also, 
- `calculationLoading.ts` - an atom indicating whether or not measure reports are being recalculated was added. 

# Testing Guidance
1. Run `npm run test` to make sure that all of the tests pass.
2. Run `npm run dev` 
3. Upload a measure bundle and click next. Create a patient. The measure highlighting for that patient should immediately render.
4. Add a resource to the patient. The measure highlighting should update to reflect that change.
5. Try out a bunch of different use cases. Add resources, delete resources, edit the patient, upload a new measure bundle. Make sure that the measure highlighting updates properly each time.
